src/parquet/schema/types.h

```diff
@@ -0,0 +1,381 @@
+// that a sequence of nodes semantically represents an array according to one
+// of these encodings (versus a struct containing an array). We should refuse
+// the temptation to guess, as they say.
+struct ListEncoding {
+  enum type {
+    ONE_LEVEL,
+    TWO_LEVEL,
+    THREE_LEVEL
+  };
+};
+
+struct DecimalMetadata {
+  int32_t scale;
+  int32_t precision;
+};
+
+// Base class for logical schema types. A type has a name, repetition level,
+// and optionally a logical type (ConvertedType in Parquet metadata parlance)
+class Node {
+ public:
+  enum type {
+    PRIMITIVE,
+    GROUP
+  };
+
+  Node(Node::type type, const std::string& name,
+      Repetition::type repetition,
+      LogicalType::type logical_type = LogicalType::NONE,
+      int id = -1) :
+      type_(type),
+      name_(name),
+      repetition_(repetition),
+      logical_type_(logical_type),
+      id_(id) {}
+
+  virtual ~Node() {}
+
+  bool is_primitive() const {
+    return type_ == Node::PRIMITIVE;
+  }
+
+  bool is_group() const {
+    return type_ == Node::GROUP;
+  }
+
+  bool is_optional() const {
+    return repetition_ == Repetition::OPTIONAL;
+  }
+
+  bool is_repeated() const {
+    return repetition_ == Repetition::REPEATED;
+  }
+
+  bool is_required() const {
+    return repetition_ == Repetition::REQUIRED;
+  }
+
+  virtual bool Equals(const Node* other) const = 0;
+
+  const std::string& name() const {
+    return name_;
+  }
+
+  Node::type node_type() const {
+    return type_;
+  }
+
+  Repetition::type repetition() const {
+    return repetition_;
+  }
+
+  LogicalType::type logical_type() const {
+    return logical_type_;
+  }
+
+  int id() const {
+    return id_;
+  }
+
+  // Node::Visitor abstract class for walking schemas with the visitor pattern
+  class Visitor {
+   public:
+    virtual ~Visitor() {}
+
+    virtual void Visit(const Node* node) = 0;
+  };
+
+  virtual void Visit(Visitor* visitor) = 0;
+
+ protected:
+  Node::type type_;
+  std::string name_;
+  Repetition::type repetition_;
+  LogicalType::type logical_type_;
+  int id_;
+
+  bool EqualsInternal(const Node* other) const;
+};
+
+// Save our breath all over the place with these typedefs
+typedef std::shared_ptr<Node> NodePtr;
+typedef std::vector<NodePtr> NodeVector;
+
+// A type that is one of the primitive Parquet storage types. In addition to
+// the other type metadata (name, repetition level, logical type), also has the
+// physical storage type and their type-specific metadata (byte width, decimal
+// parameters)
+class PrimitiveNode : public Node {
+ public:
+  PrimitiveNode(const std::string& name, Repetition::type repetition,
+      Type::type type,
+      LogicalType::type logical_type = LogicalType::NONE,
+      int id = -1) :
+      Node(Node::PRIMITIVE, name, repetition, logical_type, id),
+      physical_type_(type) {}
+
+  // FLBA ctor
+  PrimitiveNode(const std::string& name, Repetition::type repetition,
```

Comment:
Thank you for the kick in the pants here. I'm refactoring to make the constructor private with a factory methods that constructs the node (without leaking the Thrift types) from parquet metadata